web/satellite: Add bucket details in bucket view

ptdatta opened the pull request:
Change-Id: I6b932eb5209c49cf3abe9b02d59a1e5a478efc9a

Issue: #7161 

What: 
![image](https://github.com/user-attachments/assets/a4557c46-d09c-4ae3-88f0-5d8dfe49a637)


Why:

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

VitaliiShpital commented:
Hello! Thank you for contributing.

The issue here is that if you reload the page while being in the Bucket view and click `Bucket Details`, you'll be redirected to the Buckets view. I'm working on a backend change that will expose functionality to query single bucket data, so it can be used in this case.

Otherwise, it looks good!

ptdatta commented:
> Hello! Thank you for contributing.
> 
> The issue here is that if you reload the page while being in the Bucket view and click `Bucket Details`, you'll be redirected to the Buckets view. I'm working on a backend change that will expose functionality to query single bucket data, so it can be used in this case.
> 
> Otherwise, it looks good!

@VitaliiShpital On clicking the `Bucket Details`, you will redirect to Buckets view? But he users might expects to see the details of the bucket.  
Also I worked on the multinode backend I also wanted to work on the satellite. Please let me know if there is anything I can help.

VitaliiShpital commented:
@ptdatta 

Here's the piece of code responsible for that:
https://github.com/storj/storj/blob/main/web/satellite/src/components/dialogs/BucketDetailsDialog.vue#L132-L140
```
    const data = bucketsStore.state.page.buckets.find(
        (bucket: Bucket) => bucket.name === props.bucketName,
    );

    if (!data) {
        redirectToBucketsPage();

        return new Bucket();
    }
```

So, when the user reloads the page, the Pinia bucketsStore gets cleared, and the bucket details can't be found, which causes the if (!data) condition to evaluate to true.

One option to fix this is to fetch the specific bucket data from the server and display a loader while doing that, before showing the details to the user.

I'm still working on an endpoint that will allow us to do this.

Sorry for the delay. I'll let you know when I'm done

ptdatta commented:
> @ptdatta
> 
> Here's the piece of code responsible for that: https://github.com/storj/storj/blob/main/web/satellite/src/components/dialogs/BucketDetailsDialog.vue#L132-L140
> 
> ```
>     const data = bucketsStore.state.page.buckets.find(
>         (bucket: Bucket) => bucket.name === props.bucketName,
>     );
> 
>     if (!data) {
>         redirectToBucketsPage();
> 
>         return new Bucket();
>     }
> ```
> 
> So, when the user reloads the page, the Pinia `bucketsStore` gets cleared, and the bucket data can't be found, which causes the `if (!data)` condition to evaluate to `true`.
> 
> One option to fix this is to fetch the specific bucket data from the server and display a loader while doing that, before showing the details to the user.
> 
> I'm still working on an endpoint that will allow us to do this.
> 
> Sorry for the delay. I'll let you know when I'm done

Ooh now I understand. Please let me know if there is anything for me to work on the backend side.

VitaliiShpital commented:
Hey @ptdatta!

I’ve finished working on the endpoint tasks and also added an API request handler along with a Pinia module action to cover all the requirements we need to proceed.

So, if you pull the latest main branch, you’ll find a new method in the buckets store.
By using this method, you can fetch the necessary data.
https://github.com/storj/storj/blob/main/web/satellite/src/store/modules/bucketsStore.ts#L86-L90

```
    async function getSingleBucket(projectID: string, bucketName: string): Promise<Bucket> {
        const before = new Date();

        return await api.getSingle(projectID, bucketName, before);
    }
```

If you're interested, what I would do is replace the bucket's `computed` property with a `ref` inside of `BucketDetailsDialog.vue`

```
const bucket = ref<Bucket>(new Bucket());
```

Then I would try to populate it using `Vue.watch` on the `model` value set to `true`.

```
watch(model, (value) => {
    if (value) {
        const data = bucketsStore.state.page.buckets.find(bucket => bucket.name === props.bucketName);
        if (!data) {
            // do a request here and populate bucket.value
        }
    }
})
```

Please let me know if you have any questions or if you get stuck.

ptdatta commented:
> Hey @ptdatta!
> 
> I’ve finished working on the endpoint tasks and also added an API request handler along with a Pinia module action to cover all the requirements we need to proceed.
> 
> So, if you pull the latest main branch, you’ll find a new method in the buckets store. By using this method, you can fetch the necessary data. https://github.com/storj/storj/blob/main/web/satellite/src/store/modules/bucketsStore.ts#L86-L90
> 
> ```
>     async function getSingleBucket(projectID: string, bucketName: string): Promise<Bucket> {
>         const before = new Date();
> 
>         return await api.getSingle(projectID, bucketName, before);
>     }
> ```
> 
> If you're interested, what I would do is replace the bucket's `computed` property with a `ref` inside of `BucketDetailsDialog.vue`
> 
> ```
> const bucket = ref<Bucket>(new Bucket());
> ```
> 
> Then I would try to populate it using `Vue.watch` on the `model` value set to `true`.
> 
> ```
> watch(model, (value) => {
>     if (value) {
>         const data = bucketsStore.state.page.buckets.find(bucket => bucket.name === props.bucketName);
>         if (!data) {
>             // do a request here and populate bucket.value
>         }
>     }
> })
> ```
> 
> Please let me know if you have any questions or if you get stuck.

@VitaliiShpital Done ✅. I replaced the `computed` property with `ref`. Please review.

VitaliiShpital commented on web/satellite/src/components/dialogs/BucketDetailsDialog.vue:
```diff
@@ -119,26 +120,43 @@ const showRegionTag = computed<boolean>(() => {
+
+            bucket.value = fetchedBucket;
+        } catch (error) {
+            console.error('Failed to fetch bucket data', error);
```
Ok, instead of using console methods, we have a notification system. You could do something like this so the user can see what's wrong.
```
notify.notifyError(error, AnalyticsErrorEventSource.BUCKET_DETAILS_MODAL);
```

you'd have to import `useNotify` composable and add `BUCKET_DETAILS_MODAL` constant here
https://github.com/storj/storj/blob/main/web/satellite/src/utils/constants/analyticsEventNames.ts#L57-L121

something like this
```
BUCKET_DETAILS_MODAL = 'Bucket details modal',
```

VitaliiShpital commented on web/satellite/src/components/dialogs/BucketDetailsDialog.vue:
```diff
@@ -119,26 +120,43 @@ const showRegionTag = computed<boolean>(() => {
+    if (data) {
+        bucket.value = data;
+    } else {
+        try {
```
To make things a bit fancier, you could wrap this try-catch block with the existing loader handling composable `useLoading`.

It's used like this

```
const { isLoading, withLoading } = useLoading();
```

and our code would look like this

```
           withLoading(async () => {
                try {
                    bucket.value = await bucketsStore.getSingleBucket(selectedProjectID, props.bucketName);
                } catch (error) {
                    notify.notifyError(error, AnalyticsErrorEventSource.BUCKET_DETAILS_MODAL);
                    bucket.value = new Bucket();
                }
            });
```

this would allow us to set VCard's `loading` prop like this

```
<VCard :loading=isLoading>
...
</VCard>
```

By doing that, we'd add a small but stylish loading animation, which would be more noticeable in production.

VitaliiShpital reviewed (commented):
looks good! A couple of small tweaks and we're done.

VitaliiShpital reviewed (approved):
LGTM!
